jpyinterpreter-docs/src/modules/ROOT/pages/java-bytecode/java-bytecode.adoc

```diff
@@ -0,0 +1,137 @@
+
+== Generating Java bytecode instructions
+
+Java bytecode instructions are generated with the https://asm.ow2.io/[ASM library].
+Functions are created by creating a `JavaPythonClassWriter` (which is a `ClassWriter` that overrides `getCommonSuperClass` to prevent `TypeNotPresent` errors) instance,
+getting the MethodVisitor for its https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.8[functional interface method],
+adapting the method visitor with `MethodVisitorAdapters.adapt` (which sorts try blocks for you as well as give better error messages),
+and then generating the Java bytecode using that method visitor.
+It looks like this in the code:
+
+```java
+ClassWriter classWriter = new JavaPythonClassWriter(ClassWriter.COMPUTE_MAXS | ClassWriter.COMPUTE_FRAMES);
+classWriter.visit(Opcodes.V11, Modifier.PUBLIC, internalClassName, null, Type.getInternalName(Object.class),
+        new String[] { methodDescriptor.getDeclaringClassInternalName() });
+
+// ... Create fields and constructor ...
+
+MethodVisitor methodVisitor = classWriter.visitMethod(Modifier.PUBLIC,
+        methodDescriptor.getMethodName(),
+        methodDescriptor.getMethodDescriptor(),
+        null,
+        null);
+
+MethodVisitorAdapters.adapt(methodVisitor, methodDescriptor);
+
+// ... Visit parameters ...
+
+methodVisitor.visitCode();
+
+// ... Create bytecode ...
+
+methodVisitor.visitMaxs(0, 0);
+methodVisitor.visitEnd();
+```
+
+To generate a particular opcode instruction, identify what kind of instruction it is:
+
+- Instructions that either conditionally or unconditionally jump to a label are created with `visitJumpInsn`, which take the label to either conditionally or unconditionally jump to.
+
+- Instructions that load parameters or local variables are created with `visitVarInsn`, which take an `int` to identify which slot to load the parameter/local variable from.
+You can use the `LocalVariableHelper` on `StackMetadata` to obtain the slot number of parameters and local variables.
+
+- Instructions that operate on the stack are created with `visitInsn` (for instance, popping a value or adding the top two `int` on the stack). No parameter (beside the opcode to generate) is needed for these opcodes.
+
+- Instructions that call methods are created with `visitMethodInsn`, which takes.
+
+** The internal name of the declaring class.
+This can be gotten with `Type.getInternalName`.
+
+** The method name.
+
+** The method descriptor string that describe the parameter and return types of the method.
+This can be received with `Type.getMethodDescriptor(Type returnType, Type... parameterTypes)` (you can get the `Type` object of a class with `Type.getType`).
```

Comment:
In this case "you can" should probably be replaced by passive voice ("X can be retrieved"), as any personification ("we can get X") isn't really necessary here.